### ISSUE TITLE
Polytomy view performance issues

### ISSUE THREAD

**Issue by jrosindell** [OneZoom/OZtree#88]

**Comment by jrosindell** [comment]
1.) edit the node generation algorithm in some sensible way so that crazy numbers of nodes are not generated in the preload phase

2.) sprite to replace nodes that have many children until they are zoomed in

**Comment by jrosindell** [comment]
3.) Add common ancestor colours to branch styles in polytomy view - this probably means allowing the branches to be x pixels wide where x is the number of highlighted colours

**Comment by lentinj** [comment]
I was looking at (3) on the way back, hoping it was the next easiest to do. But it's trickier than I thought (which was good in a way, since I had a nice tour in the code base to figure it out).

Highlighting works by looking at the node list and adding more BezierCurves atop the original ones. But the polytomy view has PathShapes. The quick solution is to switch to using straight BezierCurves, which I think I'll do anyway since it'll be a step towards getting your curvey polytomy view, which *will* happen some day :)

However, what worries me more is, as stated before, the shape objects are very low level but also very leaky abstractions---we shouldn't have to care about what type of shape the node is, just "I want this highlighted" and the node works it out for us.

So, my proposed direction is to experiment with moving the highlighting logic to the shape objects themselves, and they manage drawing the extra lines as required. This is overkill here, however I have a hidden agenda. For #92, I'm imagining these highlights being coloured "stars" along the path. If precisely how the rendering of highlights is hidden away in the shape, this saves me spewing lots of rendering logic elsewhere.

**Comment by jrosindell** [comment]
These are great thoughts....

I agree to and like the Bezier curve solution.

In this case the branch_layout_helper and branch_layout were meant to handle the higher level questions over how to draw a connection between nodes.  However, I guess your point is that is you build a more 'intelligent' bezier curve object then you can use this object as part of nodes, leaves, and branches, signposts however you like without needing to duplicate smart line drawing logic again in the node or signpost.  I can see how this would help for non highlight related tasks which use the intelligent curves as visual embellishments instead.

**Comment by jrosindell** [comment]
One final thought on this.... we discussed before how we might have a black 'border' around white lines to help make them stand out (or a white line with different alpha even to look like a glow at the 1px scale. This would require all the wider lines that sit underneath to be a level behind the narrower lines that sit on top.  It's worth considering how this would work with more intelligent fundamental objects if those objects themselves have only one layer that they draw in.  Maybe the solution is just to use two intelligent lines at different layers but it's worth thinking about.

**Comment by lentinj** [comment]
re: black borders, my preference would be that we tell the line to have a border, and we work out how to do that at drawing time. This means if the means of rendering supports a border on a line (or we use a rect with a border to draw the line), we can make use of it.

**Comment by lentinj** [comment]
@jrosindell I see talk of a removed arrowhead that used to be on the end of lines, can I strip out this code properly now? And in general, do I have permission to remove dead code where-ever I see it?

**Comment by lentinj** [comment]
@jrosindell Also, what's the deal with draw_interior_circle()? AFAICS it puts a line-coloured circle at the end of the line, but the lines are drawn with a rounded end, and the ends are always(?) covered anyway. Is it (trying to) do something I've missed, or is it redundant?

**Comment by lentinj** [comment]
@jrosindell (I'll stop soon) https://github.com/OneZoom/OZtree/commit/1b247fa398ef67e161c2d90c2df2f2252881ed71 removes an attempt in the AT layout to highlight nodes as "concestor". What would this have achieved? I can't see anything special about the concestor nodes on live/beta.

**Comment by hyanwong** [comment]
@lentinj - did you check out concestor nodes on the AT version? E.g. `AT.html` rather than `life.html`? You'll see the text and labelling is a bit different for "concestor nodes" e.g. the one that joins chimps & humans. Apologies if I misunderstood your questions, though.

**Comment by lentinj** [comment]
I did, this isn't affecting the text / labelling, looks like there's scope for different styling for the concestor nodes. After https://github.com/OneZoom/OZtree/commit/1b247fa398ef67e161c2d90c2df2f2252881ed71 the nodes seem to look exactly the same as live, and in the process looks like I've fixed a few bugs in AT mode for beta (the search is currently completely broken, amongst others). This is probably getting sidetracked, but we might want to pull some of 3.3 into master to fix it.

**Comment by jrosindell** [comment]
@lentinj regarding deletion of dead code - yes of course I give permission for this, but please be very careful only only remove if you're certain it's dead.  There are certainly areas of dead code as well as areas of almost dead code that deal with strange edge cases. In the case of the arrows it's dead, I decided it was not necessary to show them for the visuals, they were fiddly and it was not clear from the context what direction the arrows should be facing in any case.

**Comment by jrosindell** [comment]
@lentinj In the case of the interior node circles they are needed to cover over the highlights when these are drawn in order to give a neat view of the nodes - with highlights turned off they are not necessary. Actually this is an example of code that looks dead but deals with an edge case so thanks for asking on this one - it needs to stay. 

I found that flat edges that line up 'perfectly' are sometimes rendered with a hairline gap between them - there needs to be some overlap to look good and this is the reason why I used rounded edges in some cases.

**Comment by jrosindell** [comment]
Thanks for fixing up the AT mode view

**Comment by lentinj** [comment]
I didn't say the interior node circles were dead, they're obviously still being rendered. My question is more where I should be seeing them to make sure they're still working.

**Comment by lentinj** [comment]
(1) Banning children from generating sub-nodes whenever they have more than 3 siblings seems quite effective at speeding up startup, without having a terrible affect on the results. You can occasionally see the odd artefact as loading iterates down over frames, but I don't think it's that distracting.

(2) The semi-circle of lines isn't actually that slow to render. It's the stonking number of leaves on-screen that's the problem. Replacing fake leaves and small-richness leaves with circles does no end of good. Next the ObjectPool needs upping since we hit the limit.

After all that, the next step would be to pre-generate the Leaf shapes. This would save huge amounts of time but isn't actually necessary for now, since we won't have any leaves in the OTOP theme.

**Comment by jrosindell** [comment]
This all sounds great!

> Banning children from generating sub-nodes whenever they have more than 3 siblings seems quite effective at speeding up startup, without having a terrible affect on the results. You can occasionally see the odd artefact as loading iterates down over frames, but I don't think it's that distracting.

How about setting up something a bit more bespoke than this? Before your fix it rippled down n generations in preload, regardless of children so presumably expanded nodes and passed down (n-1) until nodes were passed a 0.  You could pass down (n-floor(log_2(children number))+1) in other words it's expanding to a certain number of total descendants rather than a certain number of levels and as an approximation rounds up to the nearest whole number.  Hence in cases where there are 4 children it would ripple a little further and that wouldn't matter - similarly in cases where there are lots of groups of 3 it might not go so far. What do you think?

**Comment by jrosindell** [comment]
One comment on the highlighting look and feel.  I'm not sure about the dotted line solution here.  It is harder to see where the lines go I find it strange that the graphics are not stable to zooming in and out - the dots appear to move (presumably because of the why they are calculated).  It also doesn't scale to much larger numbers of paths (though the last point is less important)

Please could we try the nested straight lines way of visualising it as on the classic view.  Possibly with line width proportional to number of highlights required on that line?

**Comment by jrosindell** [comment]
Last point on this thread - I found a small bug with the code that decides whether to draw those circles on nodes.... if you move a circle so that it's centre is just off screen then the details in the circle are no longer drawn - really the dvar parameters of the boundary of the circle should be used for this test instead.  I think there was a related bug in the original code around the same nodes

Bug can be replicated by going to the Archaea and Eukaryotes node and moving it just off the top of the screen.  If done when zoomed in a lot the view can get stuck and can't be moved (I suspect this is because not drawing the circle also fails to register the move restrict code and therefore this code thinks movement is not allowed in any direction as nothing at all is registered on screen)

**Comment by jrosindell** [comment]
<img width="407" alt="untitled" src="https://user-images.githubusercontent.com/3009279/39960509-1243d0c8-561c-11e8-98ed-5ae4582dc583.png">
<img width="401" alt="untitled 2" src="https://user-images.githubusercontent.com/3009279/39960514-1512f7fc-561c-11e8-8eaa-f8601b655092.png">

**Comment by lentinj** [comment]
> > Banning children from generating sub-nodes whenever they have more than 3 siblings
> How about setting up something a bit more bespoke than this?

Yeah, I originally had something like what you had in mind. This is a very blunt instrument in comparison, but it seemed to do the trick so I was loath to complicate matters further for the sake of a reasonably extreme edge case. Even when it does kick in, the end result (an "unfurling") is not unpleasant. But it's not a major deal either way.

Are there known areas with small amounts of branching in the tree to experiment with?

**Comment by lentinj** [comment]
> One comment on the highlighting look and feel. I'm not sure about the dotted line solution here.
> It is harder to see where the lines go

This is a problem generally with the polytomy view, once you have significant amounts of branching, lines with any useful degree of width just results in an ugly mush. Certainly for OTOP, it's not something we're going to get away with.

> I find it strange that the graphics are not stable to zooming in and out - the dots appear to move (presumably because of the why they are calculated).

Yeah, the dot width / offsets are absolute pixels, not proportions of the line. Switching to the latter would fix it.

> It also doesn't scale to much larger numbers of paths (though the last point is less important)

Actually, I think this is one of the few ways we can scale to larger numbers of paths. The dot positions will currently start overlapping if we go over 5, but that's arbitrary and could easily be solved. OTOH, line width is not something we can arbitrarily increase.

> Please could we try the nested straight lines way of visualising it as on the classic view. Possibly with line width proportional to number of highlights required on that line?

The line type is currently very easily switchable in the layout class. My plan is to work on the OTOP theme(s) next, and in the process discover what we need for that. Then make any customisations decisions for the theme, then it will be trivial to switch the classic view back to whatever we need.

**Comment by jrosindell** [comment]
> This is a problem generally with the polytomy view, once you have significant amounts of branching, lines with any useful degree of width just results in an ugly mush. Certainly for OTOP, it's not something we're going to get away with.

How about making the thickness contingent on the number of highlights - no highlight means 1 px width, one colour to highlight means still  1px width, no colours to highlight means 3px width (with one colour in the centre). 3 colours to highlight means 5px and so on - this way it's only the highlights that stand out as required. I think that would look OK and even if later there were sprites etc. for polytomies the highlights only could be drawn over the top of them.

I think it would be nice to have this if it's a trivial switch as I'm sure it would be nicer and clearer than the dotted lines.

**Comment by lentinj** [comment]
Are we happy with the stripes we have now instead? They're also a lot more stable for zooming in/out, which solves that part of the problem. Regardless, I think we should close this and open something else if we want to choose different stylings.
